compiler/src/iree/compiler/Dialect/Flow/Conversion/MeshToFlow/MeshToFlow.cpp

```diff
@@ -0,0 +1,768 @@
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowTypes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Utils/Indexing.h"
+#include "iree/compiler/Utils/OpVisitor.h"
+#include "iree/compiler/Utils/Permutation.h"
+#include "iree/compiler/Utils/SmallVectorDenseMapInfo.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Mesh/IR/MeshOps.h"
+#include "mlir/Dialect/Mesh/Transforms/Simplifications.h"
+#include "mlir/Dialect/Mesh/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/IR/Value.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Pass/PassRegistry.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-mesh-to-flow"
+
+namespace mlir::iree_compiler::IREE::Flow {
+
+#define GEN_PASS_CLASSES
+#define GEN_PASS_REGISTRATION
+#include "iree/compiler/Dialect/Flow/Conversion/MeshToFlow/Passes.h.inc" // IWYU pragma: keep
+
+// TODO: Use from MLIR when exposed.
+static mesh::ClusterOp getMesh(Operation *op, FlatSymbolRefAttr meshSymbol,
+                               SymbolTableCollection &symbolTableCollection) {
+  return symbolTableCollection.lookupNearestSymbolFrom<mesh::ClusterOp>(
+      op, meshSymbol);
+}
+
+template <typename MeshCollectiveOp>
+static mesh::ClusterOp getMesh(MeshCollectiveOp op,
+                               SymbolTableCollection &symbolTableCollection) {
+  return getMesh(op.getOperation(), op.getMeshAttr(), symbolTableCollection);
+}
+
+static bool hasMoreThanOneMesh(Operation *op) {
+  int meshCount = 0;
+  op->walk([&meshCount](mesh::ClusterOp mesh) {
+    ++meshCount;
+    return meshCount > 1 ? WalkResult::interrupt() : WalkResult::advance();
+  });
+  return meshCount > 1;
+}
+
+static bool isDefaultChannel(mesh::ClusterOp mesh,
+                             ArrayRef<mesh::MeshAxis> meshAxes) {
+  if (mesh.getRank() != static_cast<int64_t>(meshAxes.size())) {
+    return false;
+  }
+  mesh::MeshAxis i = 0;
+  return llvm::all_of(meshAxes, [&i](mesh::MeshAxis axis) {
```

Comment:
Use `mlir::isIdentityPermutation`?

Reply:
The function in MLIR is not generic and accepts only `int64_t` for the permutation value type. I put a generic one in `Utils/Permutation.h` with a comment to make the one upstream generic.